### PR TITLE
[Modules] Create module for Microsoft.Network/dnsForwardningRulesets

### PR DESCRIPTION
**Pull request by JPEasier**
New Module `Microsoft.Network/dnsForwardningRulesets`
submodules:
- dnsForwardningRulesets/forwardingRules
- dnsForwardningRulesets/virtualNetworkLinks

Due to an underlying API issue with naming of locations I updated the deployment location in the `settings.yml` file from `West Europe` to `westeurope` - see [error messege](https://github.com/Azure/ResourceModules/actions/runs/4785801040/jobs/8508989312)

## Pipeline references

| Pipeline |
| - |
|[Network - DNS Forwarding Rulesets](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.dnsforwardingrulesets.yml)|

# Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: .github/workflows/ms.network.dnsforwardingrulesets.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        required: false
+        default: true
+      prerelease:
+        type: boolean
+        description: 'Publish prerelease module'
+        required: false
+        default: false
+  push:
+    branches:
+      - users/jpeasier/dnsForwardingRulesets
```
Comment (AlexanderSehr) [inline .github/workflows/ms.network.dnsforwardingrulesets.yml:28]: Nope

Comment (AlexanderSehr) [reply .github/workflows/ms.network.dnsforwardingrulesets.yml:28]: I guess this should be main

File: modules/Microsoft.Network/dnsForwardingRulesets/.bicep/nested_roleAssignments.bicep [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  'Storage Account Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '17d1049b-9a84-46fb-8f53-869881c3d3ab')
+  'Traffic Manager Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'a4b10055-b0c7-44c2-b00f-c7b5b3550cf7')
+  'User Access Administrator': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9')
+  'Virtual Machine Administrator Login': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '1c0163c0-47e6-4577-8991-ea5c82e286e4')
+  'Virtual Machine Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '9980e02c-c2be-4d73-94e8-173b1dc7cf3c')
+  'Virtual Machine User Login': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'fb879df8-f326-4884-b1cf-06f3ad86be52')
+  'Windows Admin Center Administrator Login': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'a6333a3e-0164-44c3-b281-7a577aff287f')
+}
+
+resource dnsResolver 'Microsoft.Network/ddosProtectionPlans@2021-08-01' existing = {
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/.bicep/nested_roleAssignments.bicep:81]: That appears to be the wrong resource. Shouldn't it be the dns forwarding rule set?

File: modules/Microsoft.Network/dnsForwardingRulesets/.test/common/dependencies.bicep [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+@description('Optional. The location to deploy resources to.')
+param location string = resourceGroup().location
+
+resource virtualNetwork 'Microsoft.Network/virtualNetworks@2022-01-01' = {
+  name: virtualNetworkName
+  location: location
+  properties: {
+    addressSpace: {
+      addressPrefixes: [
+        '10.10.100.0/24'
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/.test/common/dependencies.bicep:16]: In alignment with other modules that deploy networks (e.g., key vault common) I'd suggest to uses variables for the address spaces as well as similar address spaces. Just for consistency

File: .github/workflows/ms.network.dnsforwardingrulesets.yml (hunk 1 above)
Comment (JPEasier) [reply .github/workflows/ms.network.dnsforwardingrulesets.yml:28]: thanks, forgot to change it back

File: modules/Microsoft.Network/dnsForwardingRulesets/.bicep/nested_roleAssignments.bicep (hunk 2 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Network/dnsForwardingRulesets/.bicep/nested_roleAssignments.bicep]: Thanks @JPEasier, this breaks my heart. If you see any, please feel free to immediately fix them. In any case, I opened an issue #3162.

Comment (JPEasier) [reply modules/Microsoft.Network/dnsForwardingRulesets/.bicep/nested_roleAssignments.bicep]: i will take a look for all modules ;)

File: modules/Microsoft.Network/dnsForwardingRulesets/.test/common/deploy.test.bicep [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+module testDeployment '../../deploy.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}'
+  params: {
+    enableDefaultTelemetry: enableDefaultTelemetry
+    name: '<<namePrefix>>${serviceShort}001'
+    dnsResolverOutboundEndpointId: nestedDependencies.outputs.dnsResolverOutboundEndpointsId
+    vNetLinks: [
+      nestedDependencies.outputs.virtualNetworkId
+    ]
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/.test/common/deploy.test.bicep:54]: @JPEasier I guess we should add an RBAC test here too

File: modules/Microsoft.Network/dnsForwardingRulesets/.test/min/dependencies.bicep [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  parent: dnsResolver
+  properties: {
+    subnet: {
+      id: virtualNetwork.properties.subnets[1].id
+    }
+  }
+}
+
+@description('The resource ID of the created Virtual Network.')
+output virtualNetworkId string = virtualNetwork.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/.test/min/dependencies.bicep:75]: Suggestion use `virtualNetworkResourceId` here and in the calling module

File: modules/Microsoft.Network/dnsForwardingRulesets/.test/min/dependencies.bicep [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+@description('The resource ID of the created inbound endpoint Virtual Network Subnet.')
+output subnetResourceId_dnsIn string = virtualNetwork.properties.subnets[0].id
+
+@description('The resource ID of the created outbound endpoint Virtual Network Subnet.')
+output subnetResourceId_dnsOut string = virtualNetwork.properties.subnets[1].id
+
+@description('The resource ID of the created DNS Resolver.')
+output dnsResolverOutboundEndpointsId string = outboundEndpoints.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/.test/min/dependencies.bicep:84]: Also here, I'd suggest `dnsResolverOutboundEndpointsResourceId` instead

File: modules/Microsoft.Network/dnsForwardingRulesets/readme.md [hunk 7]
```diff
@@ -0,0 +1,322 @@
+# Network DnsForwardingRulesets `[Microsoft.Network/dnsForwardingRulesets]`
+
+This module deploys Network DnsForwardingRulesets.
+// TODO: Replace Resource and fill in description
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/readme.md:4]:
```suggestion
# DNS Forwarding Rulesets `[Microsoft.Network/dnsForwardingRulesets]`

This module deploys a DNS Forwarding Ruleset.
```

File: modules/Microsoft.Network/dnsForwardingRulesets/.test/common/deploy.test.bicep (hunk 4 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Network/dnsForwardingRulesets/.test/common/deploy.test.bicep:54]: Also, 'locks' seem to be missing

File: modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+resource dnsForwardingRulesets 'Microsoft.Network/dnsForwardingRulesets@2022-07-01' = {
+  name: name
+  location: location
+  tags: tags
+  properties: {
+    dnsResolverOutboundEndpoints: [
+      {
+        id: dnsResolverOutboundEndpointId
+      }
+    ]
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep:55]: Can there only be one

File: modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    mode: 'Incremental'
+    template: {
+      '$schema': 'https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#'
+      contentVersion: '1.0.0.0'
+      resources: []
+    }
+  }
+}
+
+resource dnsForwardingRulesets 'Microsoft.Network/dnsForwardingRulesets@2022-07-01' = {
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep:46]: I'd suggest to use `dnsForwardingRuleset` as the name as it just deploys a single resource

File: modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  properties: {
+    dnsResolverOutboundEndpoints: [
+      {
+        id: dnsResolverOutboundEndpointId
+      }
+    ]
+  }
+}
+
+module forwardingRule 'forwardingRules/deploy.bicep' = [for (forwardingRule, index) in forwardingRules: {
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep:59]:
```suggestion
module dnsForwardingRuleset_forwardingRules 'forwardingRules/deploy.bicep' = [for (forwardingRule, index) in forwardingRules: {
```

File: modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  params: {
+    dnsForwardingRulesetName: dnsForwardingRulesets.name
+    name: forwardingRule.name
+    forwardingRuleState: forwardingRule.forwardingRuleState
+    domainName: forwardingRule.domainName
+    targetDnsServers: forwardingRule.targetDnsServers
+  }
+}]
+
+module virtualNetworkLinks 'virtualNetworkLinks/deploy.bicep' = [for (vnetId, index) in vNetLinks: {
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep:70]:
```suggestion
module dnsForwardingRuleset_virtualNetworkLinks 'virtualNetworkLinks/deploy.bicep' = [for (vnetId, index) in vNetLinks: {
```

File: modules/Microsoft.Network/dnsForwardingRulesets/virtualNetworkLinks/readme.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
+# Network DnsForwardingRulesets VirtualNetworkLinks `[Microsoft.Network/dnsForwardingRulesets/virtualNetworkLinks]`
+
+This module deploys Network DnsForwardingRulesets VirtualNetworkLinks.
+// TODO: Replace Resource and fill in description
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/virtualNetworkLinks/readme.md:4]:
```suggestion
# DNS Forwarding Ruleset Virtual Network Links `[Microsoft.Network/dnsForwardingRulesets/virtualNetworkLinks]`

This module deploys Network DnsForwardingRulesets VirtualNetworkLinks.
// TODO: Replace Resource and fill in description
```

File: modules/Microsoft.Network/dnsForwardingRulesets/readme.md [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+| `location` | string | `[resourceGroup().location]` |  | Location for all resources. |
+| `lock` | string | `''` | `['', CanNotDelete, ReadOnly]` | Specify the type of lock. |
+| `roleAssignments` | array | `[]` |  | Array of role assignment objects that contain the 'roleDefinitionIdOrName' and 'principalId' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: '/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11'. |
+| `tags` | object | `{object}` |  | Tags of the resource. |
+| `vNetLinks` | array | `[]` |  | Array of virtual network links. |
+
+
+### Parameter Usage: `<ParameterPlaceholder>`
+
+// TODO: Fill in Parameter usage
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/readme.md:48]:
```suggestion
```

File: modules/Microsoft.Network/dnsForwardingRulesets/forwardingRules/readme.md [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+| :-- | :-- | :-- | :-- | :-- |
+| `enableDefaultTelemetry` | bool | `True` |  | Enable telemetry via a Globally Unique Identifier (GUID). |
+| `forwardingRuleState` | string | `'Enabled'` | `[Disabled, Enabled]` | The state of forwarding rule. |
+| `location` | string | `[resourceGroup().location]` |  | Location for all resources. |
+| `metadata` | object | `{object}` |  | Metadata attached to the forwarding rule. |
+
+
+### Parameter Usage: `<ParameterPlaceholder>`
+
+// TODO: Fill in Parameter usage
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/forwardingRules/readme.md:43]:
```suggestion
```

File: modules/Microsoft.Network/dnsForwardingRulesets/forwardingRules/deploy.bicep [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+param name string
+
+@description('Optional. Location for all resources.')
+param location string = resourceGroup().location
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableDefaultTelemetry bool = true
+
+@description('Required. Name of the parant DNS Forwarding Ruleset.')
+param dnsForwardingRulesetName string
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/forwardingRules/deploy.bicep:12]:
```suggestion
@description('Conditional. Name of the parent DNS Forwarding Ruleset. Required if the template is used in a standalone deployment.')
param dnsForwardingRulesetName string
```

Note: This requires an update of the readme too.

File: modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+resource dnsForwardingRulesets_lock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(lock)) {
+  name: '${dnsForwardingRulesets.name}-${lock}-lock'
+  properties: {
+    level: any(lock)
+    notes: lock == 'CanNotDelete' ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
+  }
+  scope: dnsForwardingRulesets
+}
+
+module dnsForwardingRulesets_roleAssignments '.bicep/nested_roleAssignments.bicep' = [for (roleAssignment, index) in roleAssignments: {
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep:87]:
```suggestion
module dnsForwardingRuleset_roleAssignments '.bicep/nested_roleAssignments.bicep' = [for (roleAssignment, index) in roleAssignments: {
```

File: modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep (hunk 8 above)
Comment (JPEasier) [reply modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep]: nope can be multible :)

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
Please add a `main.test.bicep` in the root of the `.test` folder as per the other modules. This was merged into main after you opened your PR.

File: modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep (hunk 10 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep]: Ping @JPEasier

File: modules/Microsoft.Network/dnsForwardingRulesets/virtualNetworkLinks/readme.md [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+| Parameter Name | Type | Default Value | Description |
+| :-- | :-- | :-- | :-- |
+| `enableDefaultTelemetry` | bool | `True` | Enable telemetry via a Globally Unique Identifier (GUID). |
+| `location` | string | `'global'` | The location of the PrivateDNSZone. Should be global. |
+| `name` | string | `[format('{0}-vnetlink', last(split(parameters('virtualNetworkResourceId'), '/')))]` | The name of the virtual network link. |
+
+
+### Parameter Usage: `<ParameterPlaceholder>`
+
+// TODO: Fill in Parameter usage
```
[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply modules/Microsoft.Network/dnsForwardingRulesets/virtualNetworkLinks/readme.md]: Ping @JPEasier

File: modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep (hunk 10 above)
Comment (JPEasier) [reply modules/Microsoft.Network/dnsForwardingRulesets/deploy.bicep]: file dosn't exist anymore

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**